### ISSUE TITLE
Training never achieves results of original caffe models

### ISSUE THREAD

**Issue by anatolix** [michalfaber/keras_Realtime_Multi-Person_Pose_Estimation#39]
I tried to train model with original C++ augmentation (rmpe_server) and my own python implementation (py_rmpe_server) it never train correctly.

To prove my point this is demo.py output with weights.best.h5 converted from caffe 
![canonical caffe](https://user-images.githubusercontent.com/1033375/33281664-088854e2-d3b7-11e7-8496-0eeacc088100.png)
Note it is perfect joint match and no additional unconnected points.

**Comment by anatolix** [comment]
These are training with C++ rmpe_server picture for each 10 generations
![cpp10](https://user-images.githubusercontent.com/1033375/33281765-6fbf65b0-d3b7-11e7-95f5-425292532edc.png)
![cpp20](https://user-images.githubusercontent.com/1033375/33281766-6fda8de0-d3b7-11e7-9b4d-ba64bf93c837.png)
![cpp30](https://user-images.githubusercontent.com/1033375/33281767-6ff9ee24-d3b7-11e7-9610-b53b90710474.png)
![cpp40](https://user-images.githubusercontent.com/1033375/33281768-7017db5a-d3b7-11e7-8e94-989a171c9f32.png)
![cpp50](https://user-images.githubusercontent.com/1033375/33281769-7039f14a-d3b7-11e7-8045-7c21dbb5363a.png)

there is perfect skeletons but additional points
![cpp60](https://user-images.githubusercontent.com/1033375/33281770-705566c8-d3b7-11e7-89a0-445ef91b14c2.png)

near perfect but double dots in legs of center guy on the background
![cpp70](https://user-images.githubusercontent.com/1033375/33281771-70724e78-d3b7-11e7-971f-238f2bbe8581.png)

more overfitted models later
![cpp80](https://user-images.githubusercontent.com/1033375/33281772-70922ffe-d3b7-11e7-8b3a-6361eed387bd.png)
![cpp90](https://user-images.githubusercontent.com/1033375/33281773-70b1fb18-d3b7-11e7-9a30-f3e6a4f0feef.png)
![cpp100](https://user-images.githubusercontent.com/1033375/33281774-70d10cc4-d3b7-11e7-8f57-aca4589f7b0c.png)
![cpp110](https://user-images.githubusercontent.com/1033375/33281776-70f0828e-d3b7-11e7-8c8b-821580ec2bec.png)
![cpp120](https://user-images.githubusercontent.com/1033375/33281777-7110253a-d3b7-11e7-80f5-e3f50229000a.png)

**Comment by anatolix** [comment]
These are my results with py_rmpe_server so far:
![py10](https://user-images.githubusercontent.com/1033375/33281920-f081b86a-d3b7-11e7-9004-5617bab9445f.png)
![py20](https://user-images.githubusercontent.com/1033375/33281921-f0a05b12-d3b7-11e7-9594-6b151b267a81.png)
![py30](https://user-images.githubusercontent.com/1033375/33281922-f0bc9340-d3b7-11e7-8bf6-32e0939ccce2.png)
![py40](https://user-images.githubusercontent.com/1033375/33281923-f0db63c4-d3b7-11e7-82c5-4601bccc5910.png)

**Comment by anatolix** [comment]
The questions is did this project are able to achieve results of original training at all?
It may be caught somewhere around 70th generation or may be it wasn't achieved at all?
Do original authors know some magic trick?

**Comment by anatolix** [comment]
I check val_stage6_Lx losses which inflience qualities. For Cpp augmentation best are

gen 117  for L2 loss
![cpp107](https://user-images.githubusercontent.com/1033375/33286364-32432104-d3c6-11e7-9dc0-4ef04f59599b.png)

gen 107  for L1 loss
![cpp117](https://user-images.githubusercontent.com/1033375/33286368-34a0d608-d3c6-11e7-89ca-db6911fd3c77.png)

They are almost perfect again. May be training step decay made better

**Comment by anatolix** [comment]
May be I've found problem: 
```>>> for i in range(150): print(i,step_decay(i))
0 4e-05
1 4e-05
2 4e-05
3 4e-05
...
50 4e-05
51 4e-05
52 1.3320000000000001e-05
53 1.3320000000000001e-05
...
102 1.3320000000000001e-05
103 1.3320000000000001e-05
104 4.435560000000001e-06
105 4.435560000000001e-06
...
148 4.435560000000001e-06
149 4.435560000000001e-06
```

Note: learning rate changes by large steps, due to python3 integer division. Actually we achieve small enough learning rate after generation 100.

Original code has 25 generations, each twice size as ours(meta.write_number: 121000), btw why? Did they had more images or made more agmentation? 
Learning rate changed after epoch 17 (ours will be 36) and training finished by epoch 25 (ours 50)

**Comment by anatolix** [comment]
And all of this affected by batch size, I have batch size = 20, so probably there is my problems in training

**Comment by anatolix** [comment]
Probably I need help here - should be learning rate changed with batch size?
On one thought larger batch size is larger sum of gradients, I do't remember did the use sum or mean.
On other thought larger batch mean less stochastic. 
Haven't found answer in google.
